ocaml/xapi/xapi_vtpm.ml

```diff
@@ -12,19 +12,42 @@
    GNU Lesser General Public License for more details.
  *)
 
+let assert_no_vtpm_associated ~__context vm =
+  match Db.VM.get_VTPMs ~__context ~self:vm with
+  | [] ->
+      ()
+  | vtpms ->
+      let uuids =
+        List.map (fun v -> Db.VTPM.get_uuid ~__context ~self:v) vtpms
+      in
+      raise Api_errors.(Server_error (vtpm_max_amount_reached, uuids))
+
+let assert_vm_is_halted ~__context vm =
+  match Db.VM.get_power_state ~__context ~self:vm with
+  | `Halted ->
+      ()
+  | state ->
+      let to_string = Record_util.power_to_string in
+      let expected = to_string `Halted in
+      let error_params = [Ref.string_of vm; expected; to_string state] in
+      raise Api_errors.(Server_error (vm_bad_power_state, error_params))
+
 let introduce ~__context ~uuid ~vM ~profile ~contents =
   let ref = Ref.make () in
   Db.VTPM.create ~__context ~ref ~uuid ~vM ~profile ~contents ;
   ref
 
 let create ~__context ~vM =
+  assert_no_vtpm_associated ~__context vM ;
+  assert_vm_is_halted ~__context vM ;
```

Comment:
You could use `Xapi_vm_lifecycle.assert_initial_power_state_is`.

Reply:
An alternative to this is to allow VTPM creation while the VM is running, with the understanding that it will only take effect after a VM reboot (and yes, we should make sure that reboot and shutdown+start behave the same...).